mojoshader.c

```diff
@@ -16,8554 +16,257 @@
-        if (writemask0) swiz_str[i++] = swizzle_channels[swizzle_x];
-        if (writemask1) swiz_str[i++] = swizzle_channels[swizzle_y];
-        if (writemask2) swiz_str[i++] = swizzle_channels[swizzle_z];
-        if (writemask3) swiz_str[i++] = swizzle_channels[swizzle_w];
-    } // if
-    assert(i < strsize);
-    swiz_str[i] = '\0';
-    return swiz_str;
-} // make_GLSL_swizzle_string
-
-
-static const char *make_GLSL_srcarg_string(Context *ctx, const size_t idx,
-                                           const int writemask, char *buf,
-                                           const size_t buflen)
-{
-    *buf = '\0';
-
-    if (idx >= STATICARRAYLEN(ctx->source_args))
-    {
-        fail(ctx, "Too many source args");
-        return buf;
-    } // if
-
-    const SourceArgInfo *arg = &ctx->source_args[idx];
-
-    const char *premod_str = "";
-    const char *postmod_str = "";
-    switch (arg->src_mod)
-    {
-        case SRCMOD_NEGATE:
-            premod_str = "-";
-            break;
-
-        case SRCMOD_BIASNEGATE:
-            premod_str = "-(";
-            postmod_str = " - 0.5)";
-            break;
-
-        case SRCMOD_BIAS:
-            premod_str = "(";
-            postmod_str = " - 0.5)";
-            break;
-
-        case SRCMOD_SIGNNEGATE:
-            premod_str = "-((";
-            postmod_str = " - 0.5) * 2.0)";
-            break;
-
-        case SRCMOD_SIGN:
-            premod_str = "((";
-            postmod_str = " - 0.5) * 2.0)";
-            break;
-
-        case SRCMOD_COMPLEMENT:
-            premod_str = "(1.0 - ";
-            postmod_str = ")";
-            break;
-
-        case SRCMOD_X2NEGATE:
-            premod_str = "-(";
-            postmod_str = " * 2.0)";
-            break;
-
-        case SRCMOD_X2:
-            premod_str = "(";
-            postmod_str = " * 2.0)";
-            break;
-
-        case SRCMOD_DZ:
-            fail(ctx, "SRCMOD_DZ unsupported"); return buf; // !!! FIXME
-            postmod_str = "_dz";
-            break;
-
-        case SRCMOD_DW:
-            fail(ctx, "SRCMOD_DW unsupported"); return buf; // !!! FIXME
-            postmod_str = "_dw";
-            break;
-
-        case SRCMOD_ABSNEGATE:
-            premod_str = "-abs(";
-            postmod_str = ")";
-            break;
-
-        case SRCMOD_ABS:
-            premod_str = "abs(";
-            postmod_str = ")";
-            break;
-
-        case SRCMOD_NOT:
-            premod_str = "!";
-            break;
-
-        case SRCMOD_NONE:
-        case SRCMOD_TOTAL:
-             break;  // stop compiler whining.
-    } // switch
-
-    const char *regtype_str = NULL;
-
-    if (!arg->relative)
-    {
-        regtype_str = get_GLSL_varname_in_buf(ctx, arg->regtype, arg->regnum,
-                                              (char *) alloca(64), 64);
-    } // if
-
-    const char *rel_lbracket = "";
-    char rel_offset[32] = { '\0' };
-    const char *rel_rbracket 
```

Comment:
What we can do here is move this block above the `!AT_LEAST_ONE_PROFILE` check, then change this...

```
#if !SUPPORT_PROFILE_GLSL
#define PROFILE_EMITTER_GLSL(op)
#else
#undef AT_LEAST_ONE_PROFILE
#define AT_LEAST_ONE_PROFILE 1
#define PROFILE_EMITTER_GLSL(op) emit_GLSL_##op,
#endif
```

... into this:

```
#if SUPPORT_PROFILE_GLSL
#undef AT_LEAST_ONE_PROFILE
#define AT_LEAST_ONE_PROFILE 1
#define PROFILE_EMITTER_GLSL(op) emit_GLSL_op,
PREDECLARE_PROFILE(GLSL)
#endif // SUPPORT_PROFILE_GLSL
```

Mostly trying to compress the steps needed to add a new profile. Current steps by my count:

1. Add new C file to profiles/, implement emitter
2. Add file to CMakeLists.txt
3. Add PROFILE_EMITTER_NEWPROFILE block to mojoshader.c
4. Add PREDECLARE_PROFILE block to mojoshader.c
5. Add new profile to `profiles` in mojoshader.c
6. Add to PROFILE_EMITTERS macro in mojoshader.c
7. Add to MOJOSHADER_maxShaderModel in mojoshader.c

Pretty much all of these aren't our fault, but I bet we can compress 3 and 4 into one step. If we ever remove the ARB1 group then we'll be able to get rid of step 7 as well.

Reply:
Which block are you referring to? I don't think the line linked with this comment is related to your suggestion.

Reply:
Oh crap, it got moved. Here: https://github.com/TheSpydog/MojoShader/blob/fna/mojoshader.c#L161